File: spec/jarvis/parser/interpreter_spec.rb
require "spec_helper"
require "jarvis/parser/rule"
require "jarvis/parser/interpreter"

RSpec.describe Jarvis::Parser::Interpreter do

  describe '#initialize' do
    let(:interpreter) { Jarvis::Parser::Interpreter.new }

    it 'has no default rule' do
      expect(interpreter.rules).to be_empty
    end
  end

  describe '#interpret' do
    let(:interpreter) { Jarvis::Parser::Interpreter.new }
    let(:list) { [] }
    before do
      interpreter.rules << Jarvis::Parser::Rule.new(/a/, ->(*) { list << 'a' })
      interpreter.rules << Jarvis::Parser::Rule.new(/b/, ->(*) { list << 'b' })
      interpreter.rules << Jarvis::Parser::Rule.new(/ab/, ->(*) { list << 'ab' })
    end

    describe 'on a unique match' do
      before { @result = interpreter.interpret 'b' }

      it 'executes the matching rule' do
        expect(list).to eq ['b']
      end

      it 'returns true since at least one rule matches' do
        expect(@result).to eq true
      end
    end

    describe 'without matching rules' do
      before { @result = interpreter.interpret 'c' }

      it 'executed no rules' do
        expect(list).to be_empty
      end

      it 'returns false' do
        expect(@result).to eq false
      end
    end

    describe 'with many matching rules' do
      before { @result = interpreter.interpret 'a' }

      it 'executed the first matching rule' do
        expect(list).to eq ['a']
      end

      it 'returns true' do
        expect(@result).to eq true
      end
    end
  end

end

File: lib/jarvis/parser/Interpreter.spec.ts
const _ = require('lodash');
import {expect} from 'chai';

import Dialog from '../interface/Dialog';
import { MockIO } from '../interface/IOs';
import * as Maybe from '../func/Maybe';

import Rule from './Rule';
import Interpreter from './Interpreter';

describe('Jarvis::Parser::Interpreter', () => {

  describe('#constructor', () => {
    let interpreter: Interpreter<any>;

    beforeEach(() => {
      interpreter = new Interpreter<any>();
    });

    it('has no default rule', () => {
      expect(interpreter.rules).to.be.empty;
    });
  });

  describe('#interpret', () => {
    let interpreter: Interpreter<string>;

    beforeEach(() => {
      interpreter = new Interpreter<string>();
    });
    let list: string[];

    beforeEach(() => {
      list = [];
      ['a', 'b', 'ab?'].forEach((pattern, i) => {
        interpreter.rules.push(new Rule(
          new RegExp(pattern), () => {
            list.push(pattern);
            return pattern;
        }));
      });
    });

    describe('on a unique match', () => {
      let result: Maybe.Type<string>;
      beforeEach(() => {
        result = interpreter.interpret('b');
      });

      it('executes the matching rule', () => {
        expect(list).to.eql(['b']);
      });

      it('returns true since at least one rule matches', () => {
        expect(Maybe.isDefined(result)).to.equal(true, 'Result defined');
        expect(result).to.eql('b');
      });
    });

    describe('without matching rules', () => {
      let result: Maybe.Type<string>;
      beforeEach(() => {
        result = interpreter.interpret('c');
      });

      it('executes no rules', () => {
        expect(list).to.be.empty;
      });

      it('returns false', () => {
        expect(Maybe.isUndefined(result)).to.equal(true, 'No result defined');
      });
    });

    describe('with many matching rules', () => {
      beforeEach(() => {
        interpreter.interpret('a');
      });

      it('executes the first matching rule', () => {
        expect(list).to.eql(['a']);
      });
    });
  });

});
